libkbfs/tlf_edit_history.go

```diff
@@ -0,0 +1,355 @@
+	edits TlfWriterEdits
+}
+
+func (teh *TlfEditHistory) getEditsCopyLocked() TlfWriterEdits {
+	if teh.edits == nil {
+		return nil
+	}
+	edits := make(TlfWriterEdits)
+	for user, userEdits := range teh.edits {
+		userEditsCopy := make([]TlfEdit, len(userEdits))
+		copy(userEditsCopy, userEdits)
+		edits[user] = userEditsCopy
+	}
+	return edits
+}
+
+func (teh *TlfEditHistory) getEditsCopy() TlfWriterEdits {
+	teh.lock.Lock()
+	defer teh.lock.Unlock()
+	return teh.getEditsCopyLocked()
+}
+
+func (teh *TlfEditHistory) updateRmds(rmds []ImmutableRootMetadata,
+	olderRmds []ImmutableRootMetadata) ([]ImmutableRootMetadata, error) {
+	newRmds := make([]ImmutableRootMetadata, len(olderRmds)+len(rmds))
+	copy(newRmds[:len(olderRmds)], olderRmds)
+	copy(newRmds[len(olderRmds):], rmds)
+	return newRmds, nil
+}
+
+func (teh *TlfEditHistory) calculateEditCounts(ctx context.Context,
+	rmds []ImmutableRootMetadata) (TlfWriterEdits, error) {
+	chains, err := newCRChains(ctx, teh.config, rmds, &teh.fbo.blocks, false)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set the paths on all the ops
+	_, err = chains.getPaths(ctx, &teh.fbo.blocks, teh.log, teh.fbo.nodeCache,
+		false)
+	if err != nil {
+		return nil, err
+	}
+
+	edits := make(TlfWriterEdits)
+	for _, writer := range rmds[len(rmds)-1].GetTlfHandle().ResolvedWriters() {
+		edits[writer] = nil
+	}
+
+outer:
+	for ptr, chain := range chains.byOriginal {
+		if chains.isDeleted(ptr) {
+			continue
+		}
+
+		for i, op := range chain.ops {
+			// Is this a create?
+			switch realOp := op.(type) {
+			case *createOp:
+				if realOp.renamed {
+					// Ignore renames for now.  TODO: notify about renames?
+					continue
+				}
+				if realOp.Type == Dir || realOp.Type == Sym {
+					// Ignore directories and symlinks. Because who
+					// wants notifications for those?
+					continue
+				}
+
+				// The pointer is actually the newly-referenced Block
+				for _, ref := range op.Refs() {
+					ptr = ref
+					break
+				}
+
+				// If a chain exists for the file, ignore this create.
+				if _, ok := chains.byOriginal[ptr]; ok {
+					continue
+				}
+
+				writer := op.getWriterInfo().uid
+				createdPath := op.getFinalPath().ChildPathNoPtr(realOp.NewName)
+				edits[writer] = append(edits[writer], TlfEdit{
+					Filepath:  createdPath.String(),
+					Type:      FileCreated,
+					LocalTime: op.getLocalTimestamp(),
+				})
+			case *syncOp:
+				lastOp := op
+				// Only the final writer matters, so find the last
+				// syncOp in this chain.
+				for j := len(chain.ops) - 1; j > i; j-- {
+					if syncOp, ok := chain.ops[j].(*syncOp); ok {
+						lastOp = syncOp
+						break
+					}
+				}
+
+				writer := lastOp.getWriterInfo().uid
+				t := FileModified
+				if chains.isCreated(ptr) {
+					t = FileCreated
+				}
+				edits[writer] = append(edits[writer], TlfEdit{
+					Filepath:  lastOp.getFinalPath().String(),
+					Type:      t,
+					LocalTime: lastOp.getLocalTimestamp(),
+				})
+				continue outer
```

Comment:
syncs and creates never appear in the same chain (the former is for files, the latter for dirs).  I'll add a comment.